client/state/comments/reducer.js

```diff
@@ -24,169 +24,165 @@ import {
- * @param {Number} action.siteId site identifier
- * @param {Number} action.postId post identifier
- * @param {Function|Object} updaterOrValue an updater function or a value
- * @returns {Immutable.Map} new state
- */
-function updateSpecificState( state, action, updaterOrValue ) {
-	const id = getCommentParentKey( action.siteId, action.postId );
-
-	if ( typeof updaterOrValue === 'function' ) {
-		return state.update( id, ( value ) => updaterOrValue( value ) );
-	}
-
-	return state.set( id, updaterOrValue );
-}
-
-/***
  * Comments items reducer, stores a comments items Immutable.List per siteId, postId
- * @param {Immutable.Map} state redux state
+ * @param {Object} state redux state
  * @param {Object} action redux action
- * @returns {Immutable.Map} new redux state
+ * @returns {Object} new redux state
  */
-export function items( state = Immutable.Map(), action ) {
+export function items( state = {}, action ) {
+	const { siteId, postId, commentId, skipSort } = action;
+
 	switch ( action.type ) {
 		case COMMENTS_CHANGE_STATUS:
-			return updateSpecificState( state, action, ( comments = Immutable.List() ) => {
-				updateExistingIn(
-					comments,
-					comment => comment.get( 'ID' ) === action.commentId,
-					comment => comment.set( 'status', action.status )
-				);
-			} );
+			const { status } = action;
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: map( state[ `${ siteId }-${ postId }` ], comment => {
+					if ( comment.ID !== commentId ) {
+						return comment;
+					}
+
+					return {
+						...comment,
+						status
+					};
+				} )
+			};
 		case COMMENTS_EDIT:
-			return updateSpecificState( state, action, ( comments = Immutable.List() ) => {
-				updateExistingIn(
-					comments,
-					comment => comment.get( 'ID' ) === action.commentId,
-					comment => comment.set( 'content', action.content )
-				);
-			} );
+			const { content } = action;
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: map( state[ `${ siteId }-${ postId }` ], comment => {
+					if ( comment.ID !== commentId ) {
+						return comment;
+					}
+
+					return {
+						...comment,
+						content
+					};
+				} )
+			};
 		case COMMENTS_RECEIVE:
-			// create set of ids for faster lookup for filter later
-			const newIds = Immutable.Set( action.comments.map( comment => comment.ID ) );
-
-			// we prefer freshly retrieved data, so throw away old data
-			return updateSpecificState( state, action, function( comments = Immutable.List() ) {
-				let newComments = comments.filter( ( comment ) => ! newIds.has( comment.get( 'ID' ) ) )
-											.concat( Immutable.fromJS( action.comments ) );
-
-				if ( ! action.skipSort ) {
-					newComments = newComments.sort( ( a, b ) => new Date( b.get( 'date' ) ) - new Date( a.get( 'date' ) ) );
-				}
-
-				return newComments;
-			} );
-
+			return {
+				...state,
+				[ `${ siteId }-${ postId }` ]: ! skipSort ? sortBy( action.comments, [ 'date' ] ) : action.comments
```

Comment:
I'm not sure this is correct. Does the this mean the endpoint always returns all comments for a given post?

The constant `NUMBER_OF_COMMENTS_PER_FETCH` seems to suggest otherwise.